engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala

```diff
@@ -0,0 +1,310 @@
+    case Event(ScatterCollectionSucceededResponse(jobKey, callOutputs), stateData) =>
+      handleCallSuccessful(jobKey, callOutputs, stateData)
+    case Event(AbortExecutingWorkflowCommand, stateData) => ??? // TODO: Implement!
+    case Event(_, _) => ??? // TODO: Lots of extra stuff to include here...
+  }
+
+  when(WorkflowExecutionSuccessfulState) {
+    FSM.NullFunction
+  }
+  when(WorkflowExecutionFailedState) {
+    FSM.NullFunction
+  }
+  when(WorkflowExecutionAbortedState) {
+    FSM.NullFunction
+  }
+
+  whenUnhandled {
+    case unhandledMessage =>
+      log.warning(s"$tag received an unhandled message: $unhandledMessage in state: $stateName")
+      stay
+  }
+
+  onTransition {
+    case _ -> toState if toState.terminal =>
+      log.info(s"$tag done. Shutting down.")
+      context.stop(self)
+    case fromState -> toState =>
+      log.info(s"$tag transitioning from $fromState to $toState.")
+  }
+
+  private def handleRetryableFailure(jobKey: BackendJobDescriptorKey) = {
+    // We start with index 1 for #attempts, hence invariant breaks only if jobKey.attempt > MaxRetries
+    if (jobKey.attempt <= MaxRetries) {
+      val newJobKey = jobKey.copy(attempt = jobKey.attempt + 1)
+      log.info(s"Retrying job execution for ${newJobKey.tag}")
+      /** Currently, we update the status of the old key to Preempted, and add a new entry (with the #attempts incremented by 1)
+        * to the execution store with status as NotStarted. This allows startRunnableCalls to re-execute this job */
+      val executionDiff = WorkflowExecutionDiff(Map(jobKey -> ExecutionStatus.Preempted, newJobKey -> ExecutionStatus.NotStarted))
+      val newData = stateData.mergeExecutionDiff(executionDiff)
+      stay() using startRunnableScopes(newData)
+    } else {
+      log.warning(s"Exhausted maximum number of retries for job ${jobKey.tag}. Failing.")
+      goto(WorkflowExecutionFailedState) using stateData.mergeExecutionDiff(WorkflowExecutionDiff(Map(jobKey -> ExecutionStatus.Failed)))
+    }
+  }
+
+  def handleCallSuccessful(jobKey: JobKey, outputs: CallOutputs, data: WorkflowExecutionActorData) = {
+    log.info(s"Job ${jobKey.tag} succeeded! Outputs: ${outputs.mkString("\n")}")
+    val newData = data.jobExecutionSuccess(jobKey, outputs)
+
+    if (newData.isWorkflowComplete) {
+      log.info(newData.outputsJson())
+      goto(WorkflowExecutionSuccessfulState) using newData
+    }
+    else
+      stay() using startRunnableScopes(newData)
+  }
+
+  /**
+    * Attempt to start all runnable jobs and return updated state data.  This will create a new copy
+    * of the state data including new pending persists.
+    */
+  @tailrec
+  private def startRunnableScopes(data: WorkflowExecutionActorData): WorkflowExecutionActorData = {
+    val runnableScopes = data.executionStore.runnableScopes.toList
+    val runnableCalls = runnableScopes.view collect { case k if k.scope.isInstanceOf[Call] => k } sortBy { _.index.getOrElse(-1) } map { _.tag }
```

Comment:
`runnableScopes` are actually `runnableKeys` (they're `JobKey`s), which I should rename.